---
title: "mBONITA analysis"
output: html_notebook
---

```{r}
require(UpSetR)
require(ggplot2)
require(ggsci)
require(ggpubr)
require(reshape2)
require(metap)
require(GGally)
require(ggplotify)
require(ComplexHeatmap)
```

# correlations between datasets - output from https://jupyter.circ.rochester.edu/user/mpalshik/notebooks/scratch/multiomics_networks_2022/multiomics_networks_20220609/Figure1.ipynb

```{r}
medianDF = read.csv("medianDF.csv", row.names = 1)
head(medianDF)

# add regression lines - https://www.r-bloggers.com/2016/02/multiple-regression-lines-in-ggpairs/

addRegressionLines <- function(data, mapping){
  p <- ggplot(data=data, mapping=mapping) + geom_point() + geom_smooth(method=lm)
  return(p)
}

p1 = ggpairs(medianDF, lower = list(continuous = addRegressionLines), columns = c("Transcriptomics", "Proteomics", "Phosphoproteomics"), aes(color = Condition, fill = Condition, shape = Condition)) + theme_bw() + theme(panel.grid.minor = element_line(color = 'grey', size = 0.25, linetype = 2), text = element_text(size = 16)) + scale_color_brewer(palette = 'Dark2') + scale_fill_brewer(palette = 'Dark2')  
ggsave("correlation_between_datasets_V1.svg", plot = p1, device = "svg", height = 6, width = 6)
ggsave("correlation_between_datasets_V1.png", plot = p1, device = "png", height = 6, width = 6, dpi = 1200)
ggsave("correlation_between_datasets_V1.pdf", plot = p1, device = "pdf", height = 6, width = 6, dpi = 1200)
print(p1)
```
```{r}
p1 = ggpairs(medianDF, lower = list(continuous = addRegressionLines), diag = list(continuous = "barDiag"), columns = c("Transcriptomics", "Proteomics", "Phosphoproteomics"), aes(color = Condition, fill = Condition, shape = Condition)) + theme_bw() + theme(panel.grid.minor = element_line(color = 'grey', size = 0.25, linetype = 2), text = element_text(size = 16)) + scale_color_brewer(palette = 'Dark2') + scale_fill_brewer(palette = 'Dark2')  
ggsave("correlation_between_datasets_V2.svg", plot = p1, device = "svg", height = 6, width = 6)
ggsave("correlation_between_datasets_V2.png", plot = p1, device = "png", height = 6, width = 6, dpi = 1200)
ggsave("correlation_between_datasets_V2.pdf", plot = p1, device = "pdf", height = 6, width = 6, dpi = 1200)
print(p1)
```


```{r}

reshapePvals <- function(mBonita_PA) {
  allPA <-
    reshape2::melt(
      mBonita_PA,
      id.vars = c("Pathway", "Method"),
      measure.vars =  c(
        "X1O2PlusCyA - X1O2NoCyA",
        "X1O2NoCyA - X19O2NoCyA",
        "X19O2NoCyA - X1O2PlusCyA"
      )
    )
  colnames(allPA) <-
    c("Pathway", "Method", "Contrast", "negativelog10pvalue")
  allPA$pvalue <- 10 ^ (-1 * allPA$negativelog10pvalue)
  allPA$FDR <- p.adjust(allPA$pvalue, method = "BH")
  allPA$negativelog10FDR <- (-1)*log10(allPA$FDR)
  temp <- allPA$pvalue < 0.05
  allPA <- allPA[temp,]
  return(allPA)
}
```

```{r}

#concat_PA <- read.csv("pvalues_concatenated_20220816.csv", row.names = 1, col.names= c("Pathway", "Code", "Nodes", "X1O2PlusCyA - X1O2NoCyA", "X1O2NoCyA - X19O2NoCyA", "X19O2NoCyA - X1O2PlusCyA"), check.names = FALSE)
#concat_PA$Method <- "Concatenated"
#concat_PA <- reshapePvals(concat_PA)
prot_PA <- read.csv("pvalues_proteomics_20220816.csv", row.names = 1, col.names= c("Pathway", "Code", "Nodes", "X1O2PlusCyA - X1O2NoCyA", "X1O2NoCyA - X19O2NoCyA", "X19O2NoCyA - X1O2PlusCyA"), check.names = FALSE)
prot_PA$Method <- "Proteomics"
prot_PA <- reshapePvals(prot_PA)
trans_PA <- read.csv("pvalues_transcriptomics_20221007.csv", row.names = 1, col.names= c("Pathway", "Code", "Nodes", "X1O2PlusCyA - X1O2NoCyA", "X1O2NoCyA - X19O2NoCyA", "X19O2NoCyA - X1O2PlusCyA"), check.names = FALSE) #read.csv("pvalues_transcriptomics_20220816.csv", row.names = 1, col.names= c("Pathway", "Code", "Nodes", "X1O2PlusCyA - X1O2NoCyA", "X1O2NoCyA - X19O2NoCyA", "X19O2NoCyA - X1O2PlusCyA"), check.names = FALSE)
trans_PA$Method <- "Transcriptomics"
trans_PA <- reshapePvals(trans_PA)
phosph_PA <- read.csv("pvalues_phosphoproteomics_20220816.csv", row.names = 1, col.names= c("Pathway", "Code", "Nodes", "X1O2PlusCyA - X1O2NoCyA", "X1O2NoCyA - X19O2NoCyA", "X19O2NoCyA - X1O2PlusCyA"), check.names = FALSE)
phosph_PA$Method <- "Phosphoproteomics"
phosph_PA <- reshapePvals(phosph_PA)

```

```{r}
Bonita_PA <- rbind(prot_PA, trans_PA, phosph_PA)
temp <- as.data.frame(reshape2::acast(data = Bonita_PA, Contrast ~ Pathway , value.var = 'FDR', fun.aggregate = function(x){return(metap::sumz(x)$p)}))
temp <- as.data.frame(reshape2::melt(as.matrix(temp)))
temp$Method <- "Combined"
colnames(temp) <- c("Contrast", "Pathway", "FDR", "Method")
temp$negativelog10FDR <- (-1)*log10(temp$FDR)
hist(temp$FDR)
concat_PA <- temp
Bonita_PA <- na.omit(rbind(prot_PA[,colnames(temp)], trans_PA[,colnames(temp)], phosph_PA[,colnames(temp)], temp))
```

```{r}
write.csv(Bonita_PA, "Bonita_allResults_20221010.csv")
temp <- Bonita_PA$FDR < 0.05
Bonita_PA <- Bonita_PA[temp,]
```

```{r}
ggplot(trans_PA) + geom_point(aes(x=negativelog10FDR, y=Pathway, color=Contrast)) + facet_wrap(facets = 'Method', drop=TRUE) + theme_bw()
ggsave(filename = "Bonita_trans_allResults.svg", width = 12, height = 24, units = 'in')
ggsave(filename = 'Bonita_trans_allResults.png', width = 12, height =24, units = 'in')
```
```{r}
ggplot(prot_PA) + geom_point(aes(x=negativelog10FDR, y=Pathway, color=Contrast)) + facet_wrap(facets = 'Method', drop=TRUE) + theme_bw()
ggsave(filename = "Bonita_prot_allResults.svg", width = 12, height = 24, units = 'in')
ggsave(filename = 'Bonita_prot_allResults.png', width = 12, height =24, units = 'in')
```
```{r}
ggplot(phosph_PA) + geom_point(aes(x=negativelog10FDR, y=Pathway, color=Contrast)) + facet_wrap(facets = 'Method', drop=TRUE) + theme_bw()
ggsave(filename = "Bonita_phosph_allResults.svg", width = 12, height = 24, units = 'in')
ggsave(filename = 'Bonita_phosph_allResults.png', width = 12, height =24, units = 'in')
```

```{r}
ggplot(concat_PA) + geom_point(aes(x=negativelog10FDR, y=Pathway, color=Contrast)) + facet_wrap(facets = 'Method', drop=TRUE) + theme_bw()
ggsave(filename = "Bonita_combined_pvals_allResults.svg", width = 12, height = 24, units = 'in')
ggsave(filename = 'Bonita_combined_pvals_allResults.png', width = 12, height =24, units = 'in')
```

# new mBONITA pvalues

```{r}
PAvals <- read.csv("pvalues_concatenated_20221010.csv")
PAvals$BHpvals <- p.adjust(PAvals$P.value)
hist(PAvals$BHpvals)
#PAvals <- PAvals[PAvals$BHpvals < 0.05,]
PAvals$negativeLog10_adjpvals <- (-1)*log10(PAvals$BHpvals)
#PAvals <- na.omit(PAvals)
head(PAvals)
write.csv(PAvals, "mBonita_pvalues.csv")
hist(PAvals$BHpvals)
```

```{r}
require(ggplot2)
require(ggpubr)
ggplot(PAvals) + geom_point(aes(x=negativeLog10_adjpvals, y=Pathway.Name, color=Contrast)) + theme_bw() + facet_wrap('Contrast') + xlab("-log10 (BH-corrected P value)") + ylab("Pathway Name") + xlim(0, max(PAvals$negativeLog10_adjpvals)) + theme_pubr()
```




```{r}
require(reshape2)
PAvals <- reshape2::melt(reshape2::acast(PAvals, Pathway.Name ~ Contrast, value.var = 'BHpvals'))
#reshape2::melt(reshape2::acast(PAvals, Pathway.Name ~ Contrast, value.var = 'negativeLog10_adjpvals'))
PAvals[is.na(PAvals)] <- 0
PAvals$negativeLog10_adjpvals = (-1)*log10(PAvals$value)
colnames(PAvals) <- c("Pathway Name", "Contrast", "BHpvals", "negativeLog10_adjpvals")
PAvals$Method <- 'mBONITA'

unique(Bonita_PA$Method)
```

```{r}
PAvals$Significant <- as.factor(unlist(lapply(PAvals$BHpvals, function(x){if(x>0.05){return(FALSE)}else{return(TRUE)}})))
PAvals$Contrast2 <- gsub(" vs ", "\nvs\n", PAvals$Contrast)
p1 <- ggplot(PAvals, aes(Contrast2, `Pathway Name`, fill=Significant)) + geom_tile(color = 'black', size = 0.3) + scale_x_discrete(position = "top") + theme_pubr(base_size = 16) + rotate_x_text(0) + scale_fill_brewer() + xlab("")
png(
  "mBONITA_pvalues.png",
  width = 12,
  height = 12,
  units = "in",
  res = 600
)
p1
dev.off()
svg(
  "mBONITA_pvalues.svg",
  width = 12,
  height = 12
)
p1
dev.off()
```

# Make upset plot of significant pathways from mBonita

```{r}

visList <- vector(mode = "list", length = length(unique(PAvals$Contrast)))
names(visList) <- unique(PAvals$Contrast)
for (n in unique(PAvals$Contrast)){
  visList[[n]] <- na.omit(PAvals[PAvals$Contrast %in% c(n), ])
  visList[[n]] <- as.vector(visList[[n]][visList[[n]]$BHpvals < 0.05, "Pathway Name"])
}
names(visList) <- gsub(" vs ", "\nvs ", names(visList))
#names(visList) <- gsub("O2", "O_2", names(visList))
p1 = upset(fromList(visList), order.by = c("freq", "degree"), mb.ratio = c(0.6, 0.4), empty.intersections = NULL, point.size = 5, line.size = 1 , mainbar.y.label = "Intersections", sets.x.label = "Pathways", text.scale = c(2,2, 2, 2, 1.5, 2), number.angles = 0, main.bar.color = 'black', sets.bar.color = 'black', shade.color = 'white', matrix.color = 'black') 

png(filename = paste0('upset','_mBonita_','allContrasts.png'), width = 10, height = 5, pointsize = 14, res = 600, units = 'in')
print(p1)
dev.off()

upset_mBonita_allContrasts = as.grob(p1, margin=0)
gplots::venn(visList)
```
```{r}
require(ComplexHeatmap)
lt = make_comb_mat(list_to_matrix(visList))
p1 = UpSet(lt, pt_size = unit(5, "mm"), right_annotation = upset_right_annotation(lt, add_numbers = TRUE, numbers_rot = 0,numbers_gp = gpar(cex=1)),  top_annotation = upset_top_annotation(lt, add_numbers = TRUE, numbers_rot = 0,numbers_gp = gpar(cex=1)))
upset_mBonita_allContrasts = as.grob(p1)
p1
```
# get top 5 pathways in each contrast

```{r}
for (contrast in unique(PAvals$Contrast)){
  temp_PAvals = PAvals[PAvals$Contrast == contrast,]
  temp_PAvals = temp_PAvals[order(temp_PAvals$BHpvals),]
  print(contrast)
  print(temp_PAvals[1:15,])
}
```

# Pathways only in mBonita

```{r}
unique_concat <- setdiff(PAvals$`Pathway Name`, union(concat_PA$Pathway, union(prot_PA$Pathway, union(phosph_PA$Pathway, prot_PA$Pathway))))
unique_concat <- na.omit(PAvals[PAvals$`Pathway Name` %in% unique_concat, ])
unique_concat <- unique_concat[unique_concat$negativeLog10_adjpvals > 1.303,]
write.csv(file = 'unique_mBONITA.csv', x = unique_concat)

ggplot(unique_concat) + geom_point(aes(x=negativeLog10_adjpvals, y=`Pathway Name`, color=Contrast)) + facet_wrap(facets = 'Contrast', drop=TRUE) + theme_pubr() + theme(panel.grid.major = element_line("gray",0.5), panel.grid.minor = element_line("gray",0.25))
ggsave(filename = "mBonita_uniqueResults.svg", width = 10, height = 6, units = 'in')
ggsave(filename = 'mBonita_uniqueResults.png', width = 10, height = 6, units = 'in')
```

```{r}
require(ggplot2)

ggplot(Bonita_PA) + geom_point(aes(x=negativelog10FDR, y=Pathway, color=Contrast)) + facet_wrap(facets = 'Method', drop=TRUE) + theme_bw()
ggsave(filename = "Bonita_allResults.svg", width = 12, height = 24, units = 'in')
ggsave(filename = 'Bonita_allResults.png', width = 12, height = 24, units = 'in')
```



```{r}
colnames(PAvals) <- c("Pathway", "Contrast", "negativelog10FDR", "Method", "Significant", "Contrast2")
Bonita_PA$Contrast <- gsub('X1O2NoCyA - X19O2NoCyA', '1% O2, CyA- vs 19% O2, CyA-', Bonita_PA$Contrast)
Bonita_PA$Contrast <- gsub('X1O2PlusCyA - X1O2NoCyA', '1% O2, CyA+ vs 1% O2, CyA-', Bonita_PA$Contrast)
Bonita_PA$Contrast <- gsub('X19O2NoCyA - X1O2PlusCyA', '19% O2, CyA- vs 1% O2, CyA+', Bonita_PA$Contrast)

columns <- intersect(colnames(Bonita_PA), colnames(PAvals))

mBonita_PA <- rbind(Bonita_PA[,columns], PAvals[,columns])

visList <- vector(mode = "list", length = length(unique(mBonita_PA$Method)))
names(visList) <- unique(mBonita_PA$Method)
for (n in unique(mBonita_PA$Method)){
  visList[[n]] <- na.omit(mBonita_PA[mBonita_PA$Method %in% c(n), ])
}

```

```{r}

makeMBonitaUpset <- function(contrast, visList){
  visList2 <- lapply(visList, function(x, contrast){return(as.vector(na.omit(x[x$Contrast == contrast,"Pathway"])))},contrast)
  #print(visList2)
  p1 <- upset(fromList(visList2), order.by = c("freq", "degree"), mb.ratio = c(0.55, 0.45), empty.intersections = NULL, point.size = 5, line.size = 1 , mainbar.y.label = "Intersections", sets.x.label = "Pathways", text.scale = c(2,2, 2, 2, 2, 2), number.angles = 0, main.bar.color = 'black', sets.bar.color = 'black', shade.color = 'white', matrix.color = 'black')
  print(p1)
  png(filename = paste0('upset','_mBonita_',make.names(contrast),'.png'), width = 5, height = 8, pointsize = 14, res = 600, units = 'in')
  print(p1)
  dev.off()
}

```


```{r}
for (contrast in unique(mBonita_PA$Contrast)){
  print(contrast)
 makeMBonitaUpset(contrast, visList)
}

```

# Are the pathways of interest significant?

```{r}
pathways <- as.vector(read.table("PATHWAYS IN HIF1A NETWORK KEGG IDS.txt", header=FALSE, fill=TRUE)[,1])
print(pathways)
```

```{r}
gmt <- read.table('kegg_networks_edited.gmt', header=FALSE, fill=TRUE, sep = "\t")[,1:2]
colnames(gmt) <- c('Pathway', 'KEGGID')
gmt <- na.omit(gmt[gmt$KEGGID %in% pathways, c("KEGGID", "Pathway")])
rownames(gmt) <- gmt$KEGGID
print(gmt)
```
```{r}
head(mBonita_PA)
```

```{r, eval=TRUE}
PAlist2 <- mBonita_PA[mBonita_PA$negativelog10FDR > 1.303,]
PAlist2 <- reshape2::acast(PAlist2, Pathway ~ Method ~ Contrast, value.var = 'negativelog10FDR')
PAlist2[is.na(PAlist2)] <- 0
PAlist2 <- reshape2::melt(PAlist2)
colnames(PAlist2) <- c("Pathway", "Method", "Contrast", "negativelog10FDR")
PAlist2 <- na.omit(PAlist2[PAlist2$Pathway %in% gmt$Pathway[gmt$KEGGID %in% pathways],])
PAlist2
```
```{r}
require(ggsci)
require(ggpubr)
PAlist2$Significant <- as.factor(unlist(lapply(PAlist2$negativelog10FDR, function(x){if(x<1.303){return(FALSE)}else{return(TRUE)}})))
p1 <- ggplot(PAlist2, aes(Method, Pathway, fill=Significant)) + geom_tile(color = 'black', size = 0.3) + facet_wrap(~Contrast, as.table=FALSE) + theme_pubr(base_size = 16) + rotate_x_text(90) + scale_fill_brewer() + xlab("Dataset")
png(filename = "heatmap_mBonita_HIF1pathways.png", width = 12, height = 10, pointsize = 14, res = 1200, units = 'in')
p1 
dev.off()
print(p1)
```



# Analyze node modulation score - old version

```{r}
nodeData <- read.csv("node_table.csv")
nodeData <- subset(nodeData, select = -c(X))
nodeData$modScore <- nodeData$IS*nodeData$RA
colnames(nodeData)
head(nodeData)
```


```{r}
nodeData_HIF <- nodeData[nodeData$Network %in% intersect(pathways, as.vector(na.omit(gmt$KEGGID[gmt$Pathway %in% PAlist2$Pathway[PAlist2$Significant]]))),]
dim(nodeData_HIF)
head(nodeData_HIF)
```
```{r}
nodeData_HIF$modScore <- as.vector(nodeData_HIF$modScore)
nodeData_HIF$Display.Name <- as.vector(nodeData_HIF$Display.Name)
nodeData_HIF$Dataset <- as.vector(nodeData_HIF$Dataset)


```


```{r}
p2 <- ggplot(nodeData_HIF, aes(Display.Name, Dataset,  fill=modScore)) + geom_tile(color = 'black', size = 0) + facet_wrap(~Contrast, as.table=FALSE, nrow=3) + scale_fill_gradient(low = "white", high = "red")  + xlab("Dataset") + theme_pubr(base_size = 16, x.text.angle = 90) 
png(filename = "heatmap_mBonita_HIF1pathways_modScore.png", width = 20, height = 10, pointsize = 14, res = 600, units = 'in')
p2 
dev.off()
print(p2)
```


```{r}

nodeData_HIF
colnames(nodeData_HIF)
```





```{r}
makeNodeModeHeatmaps <- function(nodeData_HIF, network, contrast) {
  require(ComplexHeatmap)
  require(circlize)
  temp <-
    na.omit(nodeData_HIF[nodeData_HIF$Network == network &
                           nodeData_HIF$Contrast == contrast, c('Display.Name', 'Contrast', 'modScore', 'Dataset')]) #B cell receptor signaling pathway
  temp <-
    reshape2::acast(temp, Display.Name ~ Dataset, value.var = 'modScore')
  temp[is.na(temp)] <- 0
  temp <- as.matrix(t(temp))
  head(temp)
  f2 = colorRamp2(seq(-1, 1, length = 3), c("blue", "white", "red"), space = "RGB")
  h1 <-
    Heatmap(
      scale(temp),
      na_col = "lightgrey",
      cluster_columns = TRUE,
      cluster_rows = FALSE,
      col = f2,
      rect_gp = gpar(col = "lightgrey", lwd = 2),
      column_title_rot = 0,
      row_title_rot = 90,
      column_title = paste0('Genes in ', network),
      row_title = contrast,
      column_title_side = "bottom",
      row_order = c(
        "Combined",
        "Transcriptomics",
        "Proteomics",
        "Phosphoproteomics"
      ),
      border = TRUE,
      heatmap_legend_param = list(
        direction = "vertical",
        title = "Node Modulation Score",
        title_position = "leftcenter-rot",
        legend_height = unit(3, 'in')
      ),
      column_names_gp = gpar(fontsize = 10)
    )
  png(
    paste0("Heatmap_", make.names(contrast), "_", network, ".png"),
    width = 12,
    height = 4,
    units = "in",
    res = 600
  )
  draw(h1, heatmap_legend_side = "left")
  dev.off()
}
```

```{r}

network = "hsa04662"
nodeData_HIF$Dataset <- as.factor(nodeData_HIF$Dataset)
levels(nodeData_HIF$Dataset) <- c("Proteomics", "Combined", "Transcriptomics", "Phosphoproteomics")
for (contrast in unique(nodeData_HIF$Contrast)){
  makeNodeModeHeatmaps(nodeData_HIF, network, contrast)
}

```

```{r}
require(ComplexHeatmap)
require(circlize)
temp <-
  na.omit(nodeData_HIF[nodeData_HIF$Network == network, c('Display.Name', 'Contrast', 'modScore', 'Dataset')]) #B cell receptor signaling pathway
temp$Contrast <- as.factor(temp$Contrast)
levels(temp$Contrast) <-
  gsub("\\-\\-", "- \u2014\n", levels(temp$Contrast))
levels(temp$Contrast) <-
  gsub("\\+\\-", "+ \u2014\n", levels(temp$Contrast))
temp$DatasetContrast <-
  paste(temp$Dataset, temp$Contrast, sep = "_")
temp <-
  reshape2::acast(temp, Display.Name ~ DatasetContrast, value.var = 'modScore')
temp[is.na(temp)] <- 0
temp <- as.matrix(t(temp))
f2 = colorRamp2(seq(-1, 1, length = 3), c("blue", "white", "red"), space = "RGB")
h1 <-
  Heatmap(
    scale(temp),
    na_col = "lightgrey",
    cluster_columns = TRUE,
    cluster_rows = FALSE,
    col = f2,
    rect_gp = gpar(col = "lightgrey", lwd = 2),
    column_title_rot = 0,
    row_title_rot = 0,
    column_title = paste0('Genes in ', network),
    column_title_side = "bottom",
    border = TRUE,
    heatmap_legend_param = list(
      direction = "horizontal",
      title = "Scaled Node Modulation Score",
      title_position = "topcenter",
      legend_width = unit(5, 'in'),
      border = 'black'
    ),
    column_names_gp = gpar(fontsize = 10),
    row_names_gp = gpar(fontsize = 10),
    row_order = order(
      gsub(
        "Combined_|Proteomics_|Transcriptomics_|Phosphoproteomics_",
        "",
        rownames(temp)
      )
    ),
    row_labels = gsub("\\_.*", "", rownames(temp)),
    row_split = gsub(
      "Combined_|Proteomics_|Transcriptomics_|Phosphoproteomics_",
      "",
      rownames(temp)
    ),
    row_title_gp = gpar(fontsize = 10), row_gap = unit(2, "mm")
  )
png(
  paste0("Heatmap_", "allContrasts_", network, ".png"),
  width = 12,
  height = 4,
  units = "in",
  res = 600
)
draw(h1, heatmap_legend_side = "top")
dev.off()
draw(h1, heatmap_legend_side = "top")
```

```{r}
order(gsub("Combined_|Proteomics_|Transcriptomics_|Phosphoproteomics_", "", rownames(temp)))
```

# Analyze node modulation score - new version

```{r}
metaNet_scores = read.csv("nodeModulation.csv", row.names = 1)
metaNet_scores$Contrast = gsub("_vs_", "\nvs\n", metaNet_scores$Contrast)
head(metaNet_scores)
```
# find the genes with highest sd in nodemodulation scores

```{r}
temp2 = aggregate(nodeModulation ~ index, metaNet_scores, sd)
colnames(temp2) <- c("index", "nodeModulation_sd")
temp2 <- temp2[order(temp2$nodeModulation, decreasing = TRUE),]
head(temp2)
highSD_genes = temp2$index[1:100]
```



```{r}
temp = t(reshape2::acast(metaNet_scores, index ~ Contrast, value.var = 'nodeModulation'))
temp = temp[,temp2$index[1:50]]
print(head(temp))
```


```{r}
require(ComplexHeatmap)
require(circlize)
f2 = colorRamp2(seq(-1, 1, length = 3), c("blue", "white", "red"), space = "RGB")
h1 <- Heatmap(scale(temp),
    col = f2,
    rect_gp = gpar(col = "lightgrey", lwd = 2),
    column_title_rot = 0,
    row_title_rot = 0,
    column_title = paste0('Genes with highly variable node modulation scores'),
    column_title_side = "bottom",
    border = TRUE,
    heatmap_legend_param = list(
      direction = "horizontal",
      title = "Scaled Node Modulation Score",
      title_position = "topcenter",
      legend_width = unit(5, 'in'),
      border = 'black'
    ),
    column_names_gp = gpar(fontsize = 10),
    row_names_gp = gpar(fontsize = 10))

png(
  "LSP1network.png",
  width = 8,
  height = 5,
  units = "in",
  res = 1200
)
draw(h1, heatmap_legend_side = "top")
dev.off()
draw(h1, heatmap_legend_side = "top")
```
```{r}
require(ComplexHeatmap)
require(circlize)
temp_binned = temp
temp_binned[temp_binned > 2000] = 2000
#f3 = colorRamp2(seq(min(temp), max(temp), length = 3), c("blue", "white", "red"), space = "RGB")
f3 = colorRamp2(seq(min(temp_binned), max(temp_binned), length = 2), c("#FFFAF0", "#4B0082"))
h1 <- Heatmap(temp_binned,
    col = f3,
    rect_gp = gpar(col = "lightgrey", lwd = 2),
    column_title_rot = 0,
    row_title_rot = 0,
    column_title = paste0('Genes with highly variable node modulation scores'),
    column_title_side = "bottom",
    border = TRUE,
    heatmap_legend_param = list(
      direction = "horizontal",
      title = "Unscaled Node Modulation Score",
      title_position = "topcenter",
      #legend_width = unit(5, 'in'),
      at = c(0, 500,1000,1500,2000),
      labels = c("0", "500", "1000", "1500", "> 2000"),
      #break_dist = c(100, 5, 4, 1),
      border = 'black'
    ),
    column_names_gp = gpar(fontsize = 10),
    row_names_gp = gpar(fontsize = 10))
png(
  "LSP1network_unscaled.png",
  width = 8,
  height = 5,
  units = "in",
  res = 1200
)
draw(h1, heatmap_legend_side = "top")
dev.off()

LSP1unscaled <- Heatmap(temp_binned,
    col = f3,
    rect_gp = gpar(col = "lightgrey", lwd = 2),
    column_title_rot = 0,
    row_title_rot = 0,
    column_title = paste0('Genes with highly variable node modulation scores'),
    column_title_side = "bottom",
    border = TRUE,
    heatmap_legend_param = list(
      direction = "horizontal",
      title = "Unscaled Node Modulation Score",
      title_position = "topcenter",
      legend_height = unit(3, 'in'),
      at = c(0, 500,1000,1500,2000),
      labels = c("0", "500", "1000", "1500", "> 2000"),
      #break_dist = c(100, 5, 4, 1),
      border = 'black'
    ),
    column_names_gp = gpar(fontsize = 10),
    row_names_gp = gpar(fontsize = 10), show_heatmap_legend = FALSE)

LSP1unscaled_legend = Legend(col_fun = f3 , direction = "horizontal",
      title = "Unscaled Node Modulation Score",
      title_position = "topcenter",
      legend_width = unit(5, 'in'),
      at = c(0, 500,1000,1500,2000),
      labels = c("0", "500", "1000", "1500", "> 2000"),
      #break_dist = c(100, 5, 4, 1),
      title_gp = gpar(fontsize=10),
      border = 'black')
draw(LSP1unscaled_legend)
```





# Make plots showing RA, evidence score, importance score and node modulation score together

```{r}
IS = read.csv("ImportanceScores.csv", row.names = 1)
RA = read.csv("RelativeAbundance.csv")
ES = read.csv("EvidenceScore.csv")
NM = read.csv("nodeModulation.csv", row.names = 1)
SD = read.csv("StandardDeviation.csv")
require(knitr)
kable(head(IS))
kable(head(RA))
kable(head(ES))
kable(head(NM))
```

```{r}
pathway = 'metaNetwork'
contrast = '19% O2, CyA-_vs_1% O2, CyA+'
genesInPathway = as.vector(IS$index[IS$Pathway %in% c(pathway)])
```

```{r}
require(plyr)
RA$index = RA$Gene
ES$index = ES$Gene
SD$index = SD$Gene
colnames(SD) = c("Gene", "Phosphoproteomics_SD", "Proteomics_SD", "Transcriptomics_SD", "index")
temp = plyr::join(IS, RA, by = c('index'), type = 'inner')
temp = plyr::join(temp, ES, by = c('index'), type = 'inner')
temp = plyr::join(temp, SD, by = c('index'), type = 'inner')
temp = plyr::join(temp, NM, by = c('index', 'Contrast'), type = 'inner')
temp = temp[temp$Pathway == pathway, c("index", "Contrast", "nodeModulation", "EvidenceScore", "ImportanceScore", "Phosphoproteomics", "Proteomics", "Transcriptomics", "Phosphoproteomics_SD", "Proteomics_SD", "Transcriptomics_SD")]
dim(temp)
colnames(temp)
head(temp)
temp = temp[temp$index %in% highSD_genes,]
#temp = temp[temp$Contrast==contrast,]
```

```{r}
require(circlize)
require(ComplexHeatmap)
f2 = colorRamp2(seq(-1, 1, length = 3), c("blue", "white", "red"), space = "RGB")

heatmapList = vector(mode = 'list', length = length(c("Phosphoproteomics", "Proteomics", "Transcriptomics", "nodeModulation", "EvidenceScore","Phosphoproteomics_SD", "Proteomics_SD", "Transcriptomics_SD", "ImportanceScore")))
temp2 = temp[,c("index", "Contrast", "EvidenceScore", "ImportanceScore", "Phosphoproteomics", "Proteomics", "Transcriptomics", "Phosphoproteomics_SD", "Proteomics_SD", "Transcriptomics_SD", "nodeModulation")]
for (i in c("Phosphoproteomics", "Proteomics", "Transcriptomics", "nodeModulation", "Phosphoproteomics_SD", "Proteomics_SD", "Transcriptomics_SD", "ImportanceScore")){
  valueVar = i
  #heatmapList[[i]] = Heatmap(as.matrix(scale(reshape2::acast(temp, index ~ Contrast, value.var = valueVar))), cluster_rows = TRUE, cluster_columns = TRUE, column_title = valueVar)
  heatmapList[[i]] = Heatmap((scale(as.matrix(temp[,i]))), col=f2, cluster_rows = FALSE, cluster_columns = FALSE, column_title = valueVar)
  draw(heatmapList[[i]], heatmap_legend_side = "top")
}

col_letters = c("1" = "pink", "2" = "purple", "3" = "blue")
heatmapList[["EvidenceScore"]] = Heatmap(t(as.matrix(temp[,"EvidenceScore"])), col=col_letters, cluster_rows = FALSE, cluster_columns = FALSE, column_title = "EvidenceScore")

```
```{r}
#ht_list = heatmapList[["nodeModulation"]] + heatmapList[["Phosphoproteomics_SD"]] + heatmapList[["Proteomics_SD"]] +  heatmapList[["Transcriptomics_SD"]] + heatmapList[["ImportanceScore"]] + heatmapList[["EvidenceScore"]] 
#heatmapList[["Phosphoproteomics"]] + heatmapList[["Proteomics"]] +  heatmapList[["Transcriptomics"]]  + heatmapList[["EvidenceScore"]]  
#draw(ht_list, cluster_rows=TRUE, row_km = 4)
  
```
```{r}
require("ggplot2")                     # Load ggplot2 package
require("GGally")                      # Load GGally package
require(ggpubr)
temp2 = temp[,c("nodeModulation", "ImportanceScore","Phosphoproteomics", "Proteomics", "Transcriptomics",  "EvidenceScore","Phosphoproteomics_SD", "Proteomics_SD", "Transcriptomics_SD")]
colnames(temp2) = c("Node Modulation\nScore", "Importance\nScore","Phosphoproteomics\nlog2 FC", "Proteomics\nlog2 FC", "Transcriptomics\nlog2 FC",  "Evidence\nScore","Phosphoproteomics\nSD", "Proteomics\nSD", "Transcriptomics\nSD")
write.csv(temp2, "processedNodeModulation_from_R.csv")
p1 = ggpairs(temp2) + theme_bw() + theme(text = element_text(size=11))
ggsave("nodeModComponents.svg", p1, dpi = 1200, width = 14, height=14)
ggsave("nodeModComponents.png", p1, dpi = 1200, width = 14, height=14)
```

```{r}
require(ggpmisc)
require(reshape2)
temp3 = melt(temp2, id.vars = "Node Modulation\nScore")
colnames(temp3) = c('nm', 'var', 'val')
p1 = ggplot(data = temp3, aes(val, nm)) + geom_point(aes(val, nm)) + facet_wrap(vars(var), nrow=2, scales= 'free') + theme_pubr() + stat_correlation(method = "spearman") + stat_poly_line() + ylab("Node Modulation Score") + xlab("") + theme(text = element_text(size=12), title = element_text(size=12))
ggsave("nodeModComponents2.svg", p1, dpi = 1200, width = 7, height=5)
ggsave("nodeModComponents2.png", p1, dpi = 1200, width = 7, height=5)
nodeModComponents2 = p1 
```
# histogram of node modulation score


```{r}
temp4 = as.data.frame(melt(as.data.frame(temp), value.vars = 'nodeModulation', id.vars = c('Contrast', 'index')))
p1 = ggplot(data = temp4) + geom_histogram(aes(value, fill = Contrast), color = 'black')+ theme_pubr(legend=c(0.7,0.75), margin=FALSE) + scale_fill_brewer('Contrast', palette = 'Blues') + ylab("Count of genes in the network") + xlab('Node modulation score')+ theme(text = element_text(size=8)) #+ scale_x_continuous(breaks=as.vector(summary(temp4$value))[1:4]) + xlim(0, as.vector(summary(temp4$value))[4])
ggsave("hist_nodeModScore.svg", p1, dpi = 1200, width = 2, height=3)
ggsave("hist_nodeModScore.png", p1, dpi = 1200, width = 2, height=3)
p1
```


```{r}
require(ggpubr)
require(ggplotify)
LSP1unscaled2 = as.grob(LSP1unscaled)
nodeModComponents2 = nodeModComponents2 + theme()
LSP1unscaled_legend = Legend(col_fun = f3 , direction = "horizontal",
      title = "Unscaled Node Modulation Score",
      title_position = "topcenter",
      legend_width = unit(5, 'in'),
      at = c(0, 500,1000,1500,2000),
      labels = c("0", "500", "1000", "1500", "> 2000"),
      #break_dist = c(100, 5, 4, 1),
      title_gp = gpar(fontsize=14),
      border = 'black')
LSP1unscaled_legend2 = LSP1unscaled_legend@grob
figure4 = ggarrange(nodeModComponents2,LSP1unscaled_legend2,LSP1unscaled2,  ncol = 1, heights = c(1,0.15,1), labels = c("A", "", "B")) 
ggsave(plot = figure4, filename = "Figure4_autoarrange.pdf", height = 10, width = 8.5)



```


# mBONITA rule analysis

```{r}
localData = read.csv('local1Data.csv')
localData = localData[,2:ncol(localData)]
```

```{r}
head(localData)
```

```{r}
colnames(localData)
```

```{r}
getIndegree = function(x){
  return(stringr::str_count(pattern = "\\[",string = x) - 1)
}
localData$indegree = unlist(lapply(as.vector(localData$equivs), getIndegree))
require(stringr)
localData$methodName = gsub("\\/.*","", localData$methodName)
localData$methodName = str_to_title(localData$methodName)
```

```{r}
require(ggplot2)
require(GGally)
require(reshape2)
ggpairs(data = localData[localData$indegree>=3,], columns = 'ruleLengths', aes(color = methodName, shape = methodName), showStrips = FALSE) 
```

```{r}
require(plyr)
localData$methodName = as.factor(localData$methodName)
mu <- as.data.frame(ddply(localData[localData$indegree>=3,], "methodName", summarise, grp.mean=mean(ruleLengths), grp.sd = median(ruleLengths)))
mu$grp.mean = unlist(lapply(mu$grp.mean, round, 1))
mu$grp.sd = unlist(lapply(mu$grp.sd, round, 1))
mu$label = paste("\U00B5 = ", mu$grp.mean, "\n", "Mdn = ", mu$grp.sd, sep = "") #
#mu$label = unlist(lapply(mu$label, trimws))
#mu$label = unlist(lapply(mu$label, str_pad, nchar(mu$label[1]), side = "right", pad = " "))
localData$methodName = gsub("Concatenated", "Combined", localData$methodName)
ERS_size_distribution = ggplot(data = localData[localData$indegree>=3,]) + geom_histogram(aes(ruleLengths), bins = 20) + theme_bw() + theme(panel.grid.minor = element_line(color = 'grey', size = 0.25, linetype = 2), text = element_text(size = 14))  + ylab("Count") + scale_x_continuous(name = "Size of ERS") + facet_wrap(~methodName, nrow=1, scales="free_y") + geom_vline(data = mu, aes(xintercept=grp.mean), color='red',linetype='dashed') + geom_label(data = mu, y = Inf, x=Inf, vjust=2, hjust=1.15, aes(label=label), color = 'red') + facet_wrap(~methodName, nrow=1, scales="free_y")
ggsave(plot = ERS_size_distribution, filename = "ERS_size_distribution.png", height = 4, width = 8, dpi = 600)
ggsave(plot = ERS_size_distribution, filename = "ERS_size_distribution.svg", height = 4, width = 8)
ggsave(plot = ERS_size_distribution, filename = "ERS_size_distribution.pdf", height = 4, width = 8)
print(ERS_size_distribution)
```



```{r}
knitr::kable(mu)
```

```{r}
ggplot(data = localData[localData$indegree>=3,]) + geom_boxplot(aes(x=ruleLengths,y=methodName))
```
```{r}
rm(localData)
```

```{r}
nodeTable = read.csv("node_table.csv")
nodeTable = nodeTable[,2:ncol(nodeTable)]
print(colnames(nodeTable))
```

```{r}
require(reshape2)
require(stringr)
nodeTable_pivot = dcast(nodeTable, Display.Name + Network  ~ Dataset, value.var = 'IS', fun.aggregate = mean)
nodeTable_pivot = nodeTable_pivot[complete.cases(nodeTable_pivot),]
colnames(nodeTable_pivot)  = str_to_title(colnames(nodeTable_pivot))
#colnames(nodeTable_pivot)  = gsub("Concatenated", "Combined", colnames(nodeTable_pivot))
head(nodeTable_pivot)
```


```{r}
require(GGally)
#wrapAt7 = function(x){return(trimws(paste(sapply(seq(1, nchar(x), 7), function(i) paste0(substring(x, i, min(i + 6, nchar(x))), '\n')), collapse='')))}
#colnames(nodeTable_pivot) = unlist(lapply(colnames(nodeTable_pivot), wrapAt7))
colnames(nodeTable_pivot) = gsub("Phosphoproteomics", "Phospho\nproteomics", colnames(nodeTable_pivot))
colnames(nodeTable_pivot) = gsub("Transcriptomics", "Transcript\nomics", colnames(nodeTable_pivot))
colnames(nodeTable_pivot) = gsub("Concatenated", "Combined", colnames(nodeTable_pivot))

IS_correlations = ggpairs(nodeTable_pivot, lower=list(continuous=addRegressionLines), columns = c(3,4,5,6), diag = list(continuous = 'barDiag'), s=0.001) + theme_bw() + theme(panel.grid.minor = element_line(color = 'grey', size = 0.25, linetype = 2), strip.text= element_text(size = 12), plot.margin=rep(unit(0,"null"),4)) + xlab("Node Importance Scores") + scale_x_continuous(breaks = c(0.5,1)) + scale_y_continuous(n.breaks = 3)
IS_correlations
```




```{r}
upset_mBonita_allContrasts2 = upset_mBonita_allContrasts 
temp = ggarrange(ERS_size_distribution, ggarrange(ggmatrix_gtable(IS_correlations), upset_mBonita_allContrasts2, ncol = 2, labels = c("B", "C"), widths = c(1.25,1)), ncol = 1, labels = c("A", ""), heights = c(1,2,1))
ggsave("Figure3_autoArranged.svg", temp, width = 8, height= 8)
ggsave("Figure3_autoArranged.pdf", temp, width = 8, height= 8, dpi = 1200)
```


















