---
title: "ActivePathways: https://CRAN.R-project.org/package=ActivePathways"
output: html_notebook
---

ActivePathways: Integrative Pathway Enrichment Analysis of Multivariate Omics Data
Framework for analysing multiple omics datasets in the context of molecular pathways, biological processes and other types of gene sets. The package uses p-value merging to combine gene- or protein-level signals, followed by ranked hypergeometric tests to determine enriched pathways and processes. This approach allows researchers to interpret a series of omics datasets in the context of known biology and gene function, and discover associations that are only apparent when several datasets are combined. The package is part of the following publication: Integrative Pathway Enrichment Analysis of Multivariate Omics Data. Paczkowska M^, Barenboim J^, Sintupisut N, Fox NS, Zhu H, Abd-Rabbo D, Mee MW, Boutros PC, PCAWG Drivers and Functional Interpretation Working Group; Reimand J, PCAWG Consortium. Nature Communications (2020) <doi:10.1038/s41467-019-13983-9>.

```{r}
require(ActivePathways)
require(readxl)
require(ggplot2)
require(stringr)
pathToReactome <- "C:\\Users\\Mukta Palshikar\\Documents\\GitHub\\moBONITA\\data\\paintomics\\" #"C:/Users/mpalshikar/Documents/moBONITA/data/paintomics"
```



```{r}
runActivePathways <-
  function(de_mrna,
           de_prot,
           de_phospho,
           GMT = 'kegg_networks.gmt') {
    mrna <-
      data.frame('adj.P.Val' = de_mrna$adj.P.Val,
                 'Identifier' = de_mrna$Identifier)
    rownames(mrna) <- de_mrna$Identifier
    prot <-
      data.frame('adj.P.Val' = de_prot$adj.P.Val,
                 'Identifier' = de_prot$Identifier)
    rownames(de_prot) <- de_prot$Identifier
    phosph <-
      data.frame('adj.P.Val' = de_phospho$adj.P.Val,
                 'Identifier' = de_phospho$Identifier)
    rownames(phosph) <- de_phospho$Identifier
    temp <- merge(x = mrna,
                  y = prot,
                  by = 'Identifier',
                  all = TRUE)
    temp <- merge(x = temp,
                  y = phosph,
                  by = 'Identifier',
                  all = TRUE)
    colnames(temp) <- c("Identifier", "mrna", "prot", "phosph")
    temp[is.na(temp)] <- 1 # as recommended in documentation
    rownames(temp) <- temp$Identifier
    temp <- temp[, colnames(temp) %in% c("mrna", "prot", "phosph")]
    temp <- as.matrix(temp)
    head(temp)
    dim(temp)
    gmt <- read.GMT(GMT)
    gmt[1:3]
    return(ActivePathways(temp, gmt, correction.method = "BH", merge.method = "Brown", background = rownames(temp)))
  }
```

# Contrast 1: 1O2NoCyA vs 1O2PlusCyA
```{r}
de_mrna <- read.csv(paste(pathToReactome,"\\contrast1\\Transcriptomics_1O2NoCyclo_vs_1O2PlusCyclo.csv", sep = ""))
de_prot <- read.csv(paste(pathToReactome,"\\contrast1\\Proteomics_1O2NoCyclo_vs_1O2PlusCyclo.csv", sep = ""))
de_phospho <- read.csv(paste(pathToReactome,"\\contrast1\\Phosphoproteomics_1O2NoCyclo_vs_1O2PlusCyclo.csv", sep = ""))
commonGenes <- union(de_mrna$Identifier, union(de_prot$Identifier, de_phospho$Identifier))
length(commonGenes)
contrast1 <- runActivePathways(de_mrna,de_prot,de_phospho)
contrast1$Contrast <- "X1O2NoCyA - X1O2PlusCyA"
```
# Contrast 2: 1O2NoCyA vs 19O2NoCyA  1O2NoCyclo_vs_19O2NoCyclo
```{r}
de_mrna <- read.csv(paste(pathToReactome,"\\contrast2\\Transcriptomics_1O2NoCyclo_vs_19O2NoCyclo.csv", sep = ""))
de_prot <- read.csv(paste(pathToReactome,"\\contrast2\\Proteomics_1O2NoCyclo_vs_19O2NoCyclo.csv", sep = ""))
de_phospho <- read.csv(paste(pathToReactome,"\\contrast2\\Phosphoproteomics_1O2NoCyclo_vs_19O2NoCyclo.csv", sep = ""))
commonGenes <- union(de_mrna$Identifier, union(de_prot$Identifier, de_phospho$Identifier))
length(commonGenes)
contrast2 <- runActivePathways(de_mrna,de_prot,de_phospho)
contrast2$Contrast <- "X1O2NoCyA vs X19O2NoCyA"
contrast2
```
# Contrast 3: 1O2PlusCyA vs 19O2NoCyA

```{r}
de_mrna <- read.csv(paste(pathToReactome,"\\contrast3\\Transcriptomics_1O2PlusCyclo_vs_19O2NoCyclo.csv", sep = ""))
de_prot <- read.csv(paste(pathToReactome,"\\contrast3\\Proteomics_1O2PlusCyclo_vs_19O2NoCyclo.csv", sep = ""))
de_phospho <- read.csv(paste(pathToReactome,"\\contrast3\\Phosphoproteomics_1O2PlusCyclo_vs_19O2NoCyclo.csv", sep = ""))
commonGenes <- union(de_mrna$Identifier, union(de_prot$Identifier, de_phospho$Identifier))
length(commonGenes)
contrast3 <- runActivePathways(de_mrna,de_prot,de_phospho)
contrast3$Contrast <- "X19O2NoCyA - X1O2PlusCyA"
contrast3
```
```{r}
allContrasts <- rbind(contrast1, contrast2, contrast3)
allContrasts$negativelog10FDR <- (-1)*log10(allContrasts$adjusted.p.val)
allContrasts$Pathway <- str_to_title(trimws(gsub("KEGG|_|\\d+$", " ", allContrasts$term.id)))
allContrasts$Pathway <- str_to_title(trimws(gsub("\\s\\-\\sHomo\\sSapiens\\s\\(Human\\)", " ", allContrasts$Pathway)))
allContrasts$Method <- "Combined"
allContrasts_temp <- apply(allContrasts,2,as.character)
write.csv(as.data.frame(allContrasts_temp), "activePathways_allResults.csv", row.names=FALSE)
allContrasts
```

```{r}
ggplot(allContrasts) + geom_point(aes(x=negativelog10FDR, y=Pathway, color=Contrast))+ theme_bw()
ggsave(filename = "ActivePathways_allResults.svg", width = 12, height = 10, units = 'in')
ggsave(filename = "ActivePathways_allResults.png", width = 12, height = 10, units = 'in')
```

